src/pyiterator.jl

```diff
@@ -29,16 +29,57 @@ end
 
     Base.done(po::PyObject, s) = ispynull(s[1])
 else
-    function Base.iterate(po::PyObject, s=_start(po))
+    """
+        PyIterator{T}(pyobject)
+
+    Wrap `pyobject::PyObject` into an iterator, that produces items of type `T`. To be more precise `convert(T, item)` is applied in each iteration. This can be useful to avoid automatic conversion of items into corresponding julia types.
+    ```jldoctest
+    julia> using PyCall
+
+    julia> l = PyObject([PyObject(1), PyObject(2)])
+    PyObject [1, 2]
+    
+    julia> piter = PyCall.PyIterator{PyAny}(l)
+    PyCall.PyIterator{PyAny}(PyObject [1, 2])
+    
+    julia> collect(piter)
+    2-element Array{Any,1}:
+     1
+     2
+    
+    julia> piter = PyCall.PyIterator{PyObject}(l)
+    PyCall.PyIterator{PyObject}(PyObject [1, 2])
+    
+    julia> collect(piter)
+    2-element Array{PyObject,1}:
+     PyObject 1
+     PyObject 2
+     ```
+    """
+    struct PyIterator{T}
+        o::PyObject
+    end
+    
+    Base.eltype(::Type{PyIterator{T}}) where T = T
+    Base.eltype(::Type{PyIterator{PyAny}}) = Any
+    Base.length(piter::PyIterator) = length(piter.o)
+
```

Comment:
Since `HasLength()` is the default `IteratorSize`, I think we need something like:

```suggestion
    Base.IteratorSize(::Type{<: PyIterator}) = Base.SizeUnknown()

```

see: https://docs.julialang.org/en/latest/manual/interfaces/

(Though it looks like `Base.iterate` has been working well in PyCall without it. I wonder why...)

In principle, we can analyse `PyObject` in the constructor (i.e., run Python code `len(o)` and see if it throws a `TypeError`. We can then add it to the second type parameter of `PyIterator` (say). But I'm not sure how useful it is. Any optimisation based on iterator traits would be irrelevant if you are calling Python.

Reply:
Ah, I get it. Code like `collect(zip(py"iter([1, 2, 3])", 1:3))` fails in the current PyCall.  Defining

```julia
Base.IteratorSize(::Type{PyObject}) = Base.SizeUnknown()
```

makes it work.  So, I think `Base.IteratorSize` should be defined for both `PyObject` and `PyIterator`.

Reply:
The advantage of analyzing the `PyObject` in the constructor to see whether it supports a length might be that there may be some functions acting on iterators that will throw an error for `SizeUnknown` iterators.  We don't want such functions to fail on Python objects if we can help it.

Reply:
I like that the current implementation is dirt simple. I think we should keep it for now and only change it, if we have real code, where it causes trouble.

Reply:
I misunderstood something, its not that complicated.